src/extension/serializer.ts

```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+    const keyPath = `${tlsPath}/key.pem`
+    const certPath = `${tlsPath}/cert.pem`
+
+    let grpcOptions = {
+      baseUrl: this.serializerServiceUrl,
+      httpVersion: '2',
+      nodeOptions: {
+        rejectUnauthorized: false, // allow self-signed certificates
+        cert: fs.readFileSync(certPath),
+        key: fs.readFileSync(keyPath),
+      },
+    } as GrpcTransportOptions
+
+    return createPromiseClient(ParserService, createGrpcTransport(grpcOptions))
+
+    // return createConnectTransport({
+    //   baseUrl: this.serializerServiceUrl,
+    //   httpVersion: '1.1',
+    // })
   }
 
-  public getNotebookDataCache(cacheId: string): NotebookData | undefined {
-    return this.notebookDataCache.get(cacheId)
+  protected async saveNotebook(
+    data: NotebookData,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Uint8Array> {
+    const marshalFrontmatter = this.lifecycleIdentity === RunmeIdentity.ALL
+
+    const notebook = ConnectSerializer.marshalNotebook(data, { marshalFrontmatter })
+
+    if (marshalFrontmatter) {
+      data.metadata ??= {}
+      data.metadata[RUNME_FRONTMATTER_PARSED] = notebook.frontmatter
+    }
+
+    const cacheId = GrpcSerializerBase.getDocumentCacheId(data.metadata)
+    this.notebookDataCache.set(cacheId as string, data)
+
+    const serialRequest = new es_proto.SerializeRequest({ notebook })
+
+    const cacheOutputs = this.cacheNotebookOutputs(notebook, cacheId)
+    const request = this.client.serialize(serialRequest)
+
+    // run in parallel
+    const [serialResult] = await Promise.all([request, cacheOutputs])
+
+    if (cacheId) {
+      await this.saveNotebookOutputsByCacheId(cacheId)
+    }
+
+    if (serialResult.result === undefined) {
+      throw new Error('serialization of notebook failed')
+    }
+
+    return serialResult.result
+  }
+
+  protected async reviveNotebook(
+    content: Uint8Array,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Serializer.Notebook> {
+    const identity = this.lifecycleIdentity
+    // const markdown = Buffer.from(content).toString('utf8')
+    const deserializeRequest = new es_proto.DeserializeRequest({
+      source: content,
+      options: { identity },
+    })
+    let res
+    try {
+      res = await this.client.deserialize(deserializeRequest)
+    } catch (e: any) {
+      if (e.name === 'ConnectError') {
+        e.message = `Unable to connect to the serializer service at ${this.serializerServiceUrl}`
+      }
+      log.error('Error in reviveNotebook  ', e as any)
+      throw e
+    }
+    const notebook = res.notebook
+
+    if (!notebook) {
+      return this.printCell('⚠️ __Error__: no cells found!')
+    }
+    return notebook as any as Serializer.Notebook // ugly cast :(
+  }
+
+  protected applyIdentity(data: es_proto.Notebook): es_proto.Notebook {
```

Comment:
For applyIdentity, switchLifecycleIdentity, and other methods that are duplicated in the ConnectSerializer; if you converted the es_proto to a timostamp proto then you could just call  a function that was also used by the gRPCSerializer

e.g.

applyIdentityESProto(data: es_proto.Notebook) {
   ts_proto = convert(data)
   new_ts_proto = applyIdentityTSProto(ts_proto)
   es_proto = convert_ts_to_es(data)

   return es_proto
}

Reply:
it seems like it would be this way or a refactor to replace ts_proto with es_proto as in your next comment.  let me know which approach you feel is the best spend of time here.

Reply:
This comment is no longer relevant now that we going with a single class using just es_proto. I wrote this before it occurred to me thats what we should do.